packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -1432,38 +1363,136 @@ class TextPainter {
+  //    choice we makes in 5. allows us to still check (x - 1) in case x points
+  //    to a multi-code-unit character.
+  //
+  // 5. (x, downstream || upstream), where x points to a multi-code-unit
+  //    character. There's no perfect caret placement in this case. Here we chose
+  //    to draw the caret at the location that makes the most sense when the
+  //    user wants to backspace (which also means it's left-arrow-key-biased):
+  //
+  //     * downstream: show the caret at the leading edge of the character only if
+  //       x points to the start of the grapheme. Otherwise show the caret at the
+  //       leading edge of the next logical character.
+  //     * upstream: show the caret at the trailing edge of the previous character
+  //       only if x points to the start of the grapheme. Otherwise place the
+  //       caret at the trailing edge of the character.
+  _LineCaretMetrics? _computeCaretMetrics(TextPosition position) {
     assert(_debugAssertTextLayoutIsValid);
     assert(!_debugNeedsRelayout);
+
     final _TextPainterLayoutCacheWithOffset cachedLayout = _layoutCache!;
-    if (position == cachedLayout._previousCaretPosition) {
-      return _caretMetrics;
+    // If nothing is laid out, top start is the only reasonable place to place
+    // the cursor.
+    // The HTML renderer reports numberOfLines == 1 when the text is empty:
+    // https://github.com/flutter/flutter/issues/143331
+    if (cachedLayout.paragraph.numberOfLines < 1 || plainText.isEmpty) {
+      // TODO(LongCatIsLooong): assert when an invalid position is given.
+      return null;
     }
-    final int offset = position.offset;
-    final _CaretMetrics? metrics = switch (position.affinity) {
-      TextAffinity.upstream => _getMetricsFromUpstream(offset) ?? _getMetricsFromDownstream(offset),
-      TextAffinity.downstream => _getMetricsFromDownstream(offset) ?? _getMetricsFromUpstream(offset),
+
+    final (int offset, bool anchorToLeadingEdge) = switch (position) {
+      TextPosition(offset: 0) => (0, true), // As a special case, always anchor to the leading edge of the first grapheme regardless of the affinity.
+      TextPosition(:final int offset, affinity: TextAffinity.downstream) => (offset, true),
+      TextPosition(:final int offset, affinity: TextAffinity.upstream) when _isNewlineAtOffset(offset - 1) => (offset, true),
+      TextPosition(:final int offset, affinity: TextAffinity.upstream) => (offset - 1, false)
     };
-    // Cache the input parameters to prevent repeat work later.
-    cachedLayout._previousCaretPosition = position;
-    return _caretMetrics = metrics ?? const _EmptyLineCaretMetrics(lineVerticalOffset: 0);
+
+    final int caretPositionCacheKey = anchorToLeadingEdge ? offset : -offset - 1;
+    if (caretPositionCacheKey == cachedLayout._previousCaretPositionKey) {
+      return _caretMetrics;
+    }
+
+    final ui.GlyphInfo? glyphInfo = cachedLayout.paragraph.getGlyphInfoAt(offset);
```

Comment:
Can you explain the secret to removing the logarithmic search? Are you able to get the bounds of the glyph using this getGlyphInfoAt method instead? What does it do internally if so?

Reply:
`getBoxForSelection` doesn't include a multi-code-unit character if the input selection doesn't include the full range of it. Alternatively the same can be done using the characters package.